### PR TITLE
Add LR Scheduler support to candidate

### PR DESCRIPTION
**Pull request by titu1994**
# LR Scheduler Support

- Adds all of NeMo's schedulers to namespace `nemo.core.optim.lr_scheduler` - can be accessed via `nemo.core.optim` directly.
- Adds `prepare_scheduler` method which consumes a configuration dictionary for the scheduler and generates a LR scheduler contained in a dictionary which is required by PTL.
- Refactors arguments to fall under the name space `nemo.utils.arguments`.

# Usage

### Construct a config dictionary for optimizer and scheduler support

```python
    # Setup optimizer and scheduler
    scheduler_args = {
        'monitor': 'val_loss',  # pytorch lightning requires this value only when using ReduceLROnPlateau; ignored otherwise.
        'warmup_ratio': args.warmup_ratio,
        'warmup_steps': args.warmup_steps,
        'min_lr': args.min_lr,
        'last_epoch': args.last_epoch,
    }

    if args.max_steps is None:
        iters_per_batch = args.max_epochs / float(args.gpus * args.num_nodes * args.accumulate_grad_batches)
        scheduler_args['iters_per_batch '] = iters_per_batch 
    else:
        scheduler_args['max_steps'] = args.max_steps
```

### Feed this config to `ModelPT.setup_optimization()` as follows : 

```python
    asr_model.setup_optimization(
        optim_params={
            'optimizer': args.optimizer,
            'lr': args.lr,
            'opt_args': args.opt_args,
            'scheduler': CosineAnnealing,
            'scheduler_args': scheduler_args,
        }
    )
```

### Override `setup_optimization()` as follows : 

```python
    def setup_optimization(self, optim_params: Optional[Dict] = None) -> torch.optim.Optimizer:
        self.__optimizer = super().setup_optimization(optim_params)
        self.__scheduler = prepare_lr_scheduler(optimizer=self.__optimizer, scheduler_config=optim_params,
                                                train_dataloader=self.__train_dl)
```

### Override `configure_optimizers()` as follows : [Note, those extra lists around optimizer and scheduler is important, pytorch lightning expects it in this format]

```python
    def configure_optimizers(self):
        if self.__scheduler is None:
            return self.__optimizer
        else:
            return [self.__optimizer], [self.__scheduler]
```

### CODE REVIEW

File: examples/asr/speech_to_text.py [hunk 1]
```diff
@@ -51,7 +51,31 @@ def main(args):
...
     asr_model.setup_validation_data(model_config['AudioToTextDataLayer_eval'])
-    asr_model.setup_optimization(optim_params={'optimizer': args.optimizer, 'lr': args.lr, 'opt_args': args.opt_args})
+
+    # Setup optimizer and scheduler
+    scheduler_args = {
+        'monitor': 'val_loss',  # pytorch lightning requires this value
+        'warmup_ratio': args.warmup_ratio,
+        'warmup_steps': args.warmup_steps,
+        'min_lr': args.min_lr,
+        'last_epoch': args.last_epoch,
```
Comment (yzhang123) [inline examples/asr/speech_to_text.py:61]: what does monitor and last_epoch do?

File: examples/asr/speech_to_text.py [hunk 2]
```diff
@@ -65,6 +89,7 @@ def main(args):
     parser = pl.Trainer.add_argparse_args(parser)
     parser = add_asr_args(parser)
     parser = add_optimizer_args(parser)
+    parser = add_scheduler_args(parser)
```
Comment (yzhang123) [inline examples/asr/speech_to_text.py:92]: why are you adding it to plTrainer parser?

File: nemo/core/optim/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+
+        super().__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        if not self._get_lr_called_within_step:
+            warnings.warn(
+                "To get the last learning rate computed by the scheduler, " "please use `get_last_lr()`.", UserWarning
+            )
+
+        step = self.last_epoch
```
Comment (yzhang123) [inline nemo/core/optim/lr_scheduler.py:72]: last_epoch is the current step?!

File: examples/asr/speech_to_text.py [hunk 4]
```diff
@@ -51,7 +51,31 @@ def main(args):
...
+    scheduler_args = {
+        'monitor': 'val_loss',  # pytorch lightning requires this value
+        'warmup_ratio': args.warmup_ratio,
+        'warmup_steps': args.warmup_steps,
+        'min_lr': args.min_lr,
+        'last_epoch': args.last_epoch,
+    }
+
+    if args.max_steps is None:
+        iters_per_sample = args.max_epochs / float(args.gpus * args.num_nodes * args.accumulate_grad_batches)
```
Comment (yzhang123) [inline examples/asr/speech_to_text.py:65]: shouldnt batch size be also in the denominator? 
this term is so weird, since it only makes sense when multiplied with len(dataset) which is done later. but ok

File: examples/asr/speech_to_text.py (hunk 2 above)
Comment (titu1994) [reply examples/asr/speech_to_text.py:92]: This is to add the lr scheduling parameters to argparse. Almost all of our schedulers depend on 4-5 common parameters so this will include them. Special ones will still need to add the args though.

File: nemo/core/optim/lr_scheduler.py (hunk 3 above)
Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:72]: Yes, pytorch naming is weird. "last_epoch" is the global index of the data loader and is equivalent to iteration step for us.

File: examples/asr/speech_to_text.py (hunk 4 above)
Comment (titu1994) [reply examples/asr/speech_to_text.py]: Ok so batch size is not here, but deferred to the train_dataloader. Renamed it to `iters_per_batch` instead.

File: nemo/utils/arguments.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+def add_scheduler_args(parent_parser: ArgumentParser) -> ArgumentParser:
+    """Extends existing argparse with default LR scheduler args.
+
+    Args:
+        parent_parser (ArgumentParser): Custom CLI parser that will be extended.
+
+    Returns:
+        ArgumentParser: Parser extended by LR Scheduler arguments.
+    """
```
Comment (blisc) [inline nemo/utils/arguments.py:67]: How come add_scheduler_args doesn't have the nice `optimize` `default_lr`, and `default_opt_args` that add_optimizer_args has to add default values?

Comment (titu1994) [reply nemo/utils/arguments.py:67]: Those defaults are part of the optimizer not the scheduler, so it's included in the add optimizer args. But you are right, default scheduler args should take in a static dict / config object and set the defaults in argparse so that we have production ready schedulers. 

I'll submit another pr, this one was merged overnight.

File: nemo/core/optim/lr_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    # Compute effective max_steps if iters_per_batch is provided
+    if 'iters_per_batch' in scheduler_args:
+        if train_dataloader is None:
+            raise ValueError(
+                'As `iters_per_batch` is provided, it is required to pass the train dataloader in order '
+                'to compute effective maximum number of steps'
+            )
+
+        iters_per_batch = scheduler_args.pop('iters_per_batch')
+        num_samples = len(train_dataloader.dataset)
```
Comment (blisc) [inline nemo/core/optim/lr_scheduler.py:382]: +@redoctopus 
Is this going to crash Iterable datasets? Since they implement iter() and not getitem()/len()

**Comment by blisc** [comment]
It also be real neat to have an function that does this:
```
parser = add_optimizer_args(parser, defaults=...)
parser = add_scheduler_args(parser, defaults=...)
args = parser.parse_args()
```
and
```
setup_optimizer(args)
```

instead of
```
optim_params = {"optimizer": args.optimizers, ...}
setup_optimizer(optim_params)
```

File: nemo/core/optim/lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+        batch_size = train_dataloader.batch_size
+        max_steps = round(num_samples * iters_per_batch / float(batch_size))
```
Comment (blisc) [inline nemo/core/optim/lr_scheduler.py:384]: Can you explain this part to me?

Shouldn't it be
```
steps_in_epoch = len(train_data_layer) / (args.batch_size * args.iter_per_step * nf.world_size)
max_steps = steps_in_epoch * num_epochs
```

File: nemo/core/optim/lr_scheduler.py (hunk 6 above)
Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:382]: I checked before, https://github.com/NVIDIA/NeMo/blob/master/nemo/collections/asr/data_layer.py#L583, but it seems TarredAudioTextLayer does implement Len(). @redoctopus can you confirm this works? And after discussing with @yzhang123, I found that most IterableDataset require you to pass max_steps anyway instead of allowing the computation of effective max steps (since they may not implement __len__)

**Comment by titu1994** [comment]
@blisc I actually proposed that with a solution of passing `vars(args)` as the input to the setup function, but @VahidooX commented that it hides away a lot of clarity in what exactly we are passing to the setup functions. I believe a config object will resolve this, so we can switch to that as soon as it is stable and implement in candidate.

File: nemo/core/optim/lr_scheduler.py (hunk 7 above)
Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:384]: We don't have the notion of drop last epoch right now, though we can add it in the future. Round() is basically the switch between ceil and floor for us without the flag for the moment. 

Iters per step is not accumulate grad batches, nf.world size is gpus * num_nodes. Right now I am expecting user to precompute the epochs / denominator (disregarding batch size in denom and number of samples in numerator) and then I inject those values from the train data layer. Once we are capable of passing config objects, we can hide this internal computation.
